content/articles/automating-jobs-scheduling-with-django_cron-in-python-application/index.md

```diff
@@ -0,0 +1,394 @@
+      'rest_framework.permissions.AllowAny',
+  ],
+}
+```
+
+#### Making models for the endpoint
+In this API, what is actually needed from the external endpoint is the `ids` of the external news data. For this reason, we will make the model with just two properties which are the `id` and the `datetime` fetched.
+
+It is important to make the primary key of the database from the fetched data `id`. This is because it will be referenced whenever you are to get the latest news from the external endpoint.
+
+Add the snippet below into your `models.py` of the `news` app.
+
+```python
+from django.db import models
+from datetime import datetime
+
+class HackerNewsID(models.Model):
+    hackernews = models.BigIntegerField(unique=True, primary_key=True)
+    fetched_at = models.DateTimeField(default=datetime.now())
+
+    def save(self, *args, **kwargs):
+        self.id = self.hackernews # replacing the id(primary key) as the hackernews id
+        super(HackerNewsID, self).save(*args, **kwargs)
+
+    def __str__(self):
+        return str(self.hackernews)
+```
+
+- Note: The `save` method under the class `HackerNewsID` will be called whenever data is been saved. This only formats the primary key to the `id` fetched.
+
+#### Making cron jobs
+Here, we will make the cron job class. This class is required by the `django_cron` library to run every minute(s) as configured in the code below.
+
+The code snippet below describes the cron job class. This goes inside the `cron.py` file in the `news` app directory.
+
+```python
+from django_cron import CronJobBase, Schedule
+import requests
+from .models import HackerNewsID
+
+class MyCronJob(CronJobBase):
+    RUN_EVERY_MINS = 5 # every 5 minutes
+    RETRY_AFTER_FAILURE_MINS = 1
+    schedule = Schedule(run_every_mins=RUN_EVERY_MINS, retry_after_failure_mins=RETRY_AFTER_FAILURE_MINS)
+    code = 'news.my_cron_job'    # a unique code
+
+    def do(self):
+
+        NEWS_URL = 'https://hacker-news.firebaseio.com/v0/topstories.json?print=pretty' # external endpoint that returns list of ids
+
+        headers = {'user-agent': 'quickcheck/0.0.1'}
+        response = requests.get(NEWS_URL, headers=headers)
+
+        result = response.text.split(',')[1:len(response.text.split(','))-2] # in order to trim the last element
+        last = response.text.split(',')[-1] # got this from API " 499287535 ] /n" --> reshaped to that below
+        result.insert(len(result), last.strip().split()[0]) # "499287535"
+
+        news = 400 # 100 downward/latest
+        res = [int(id.strip()) for id in result[news+1:news+6]] # list comprehension
+
+        for id in res:
+            news_id = HackerNewsID(hackernews=id)
+            news_id.save()
+```
+- Note:  Run the command below to add the latest hackernews id to database. This will execute the cron job class(es) by looping through the response from API data, and then call `save()` method on each.
+
+
+```bash
+python manage.py runcrons
+```
```

Comment:
Using the latest django versions eg. `4.0.1` cause `django_cron` to throw the following error.

```bash
ImportError: cannot import name 'ugettext_lazy' from 'django.utils.translation'

.../python3.8/site-packages/django_cron/admin.py
```
This problem comes from the library, not your code. Please mention about it in your article. ie. including the package version

Reply:
resolved.